src/main/java/org/webshar/hrms/service/LeaveApplicationService.java

```diff
@@ -197,36 +197,24 @@ private int getNoOfDaysLeaveApplied(final Date startDate, final Date endDate)
     return (int) ChronoUnit.DAYS.between(ldStartDate, ldEndDate) + 1;
   }
 
-  private boolean isStartAndEndDateAreOverlappingWithExistingRecordForGivenEmployeeAndLeaveTypeId(
+  private void isStartAndEndDateAreOverlappingWithExistingRecordForGivenEmployeeAndLeaveTypeId(
       final Long employeeId, final Long leaveTypeId, final Date startDate, final Date endDate)
       throws EntityAlreadyExistsException
   {
-    List<LeaveApplication> leaveApplicationsList = leaveApplicationRepository
-        .findAllByEmployeeIdAndLeaveTypeIdAndStartDateLessThanEqualAndEndDateGreaterThanEqual(
-            employeeId, leaveTypeId, startDate, endDate);
-    //Need to improve this logic
-    leaveApplicationsList.addAll(
-        leaveApplicationRepository
-            .findAllByEmployeeIdAndLeaveTypeIdAndStartDateGreaterThanEqualAndEndDateLessThanEqual(
-                employeeId, leaveTypeId, startDate, endDate));
-    //Need to optimize this currently just ignore this implementation
-//    leaveApplicationsList.addAll(
-//        leaveApplicationRepository
-//            .findAllByEmployeeIdAndLeaveTypeIdAndStartDateGreaterThanEqualAndEndDateGreaterThanEqual(
-//                employeeId, leaveTypeId, startDate, endDate));
-//    leaveApplicationsList.addAll(
-//        leaveApplicationRepository
-//            .findAllByEmployeeIdAndLeaveTypeIdAndStartDateLessThanEqualAndEndDateLessThanEqual(
-//                employeeId, leaveTypeId, startDate, endDate));
-
-    if (!leaveApplicationsList.isEmpty())
+    List<LeaveApplication> leaveApplicationList = leaveApplicationRepository.findAllByStartDateGreaterThanEqualAndStartDateLessThanEqual(startDate,endDate);
+
+    leaveApplicationList.addAll(leaveApplicationRepository.findAllByEndDateGreaterThanEqualAndEndDateLessThanEqual(startDate,endDate));
+    leaveApplicationList.addAll(leaveApplicationRepository.findAllByStartDateLessThanEqualAndEndDateGreaterThanEqual(startDate,endDate));
+
+    Optional<LeaveApplication>  leaveApplicationResult = leaveApplicationList
+        .stream()
+        .filter(leaveApplication -> (leaveApplication.getEmployee().getId() == employeeId) && (leaveApplication.getLeaveType().getId() == leaveTypeId))
+        .findFirst();
+
+    if (leaveApplicationResult.isPresent())
```

Comment:
Need to improve leave overlapping validation logic.
Explore whether this operation can be performed by a single read/query. 
Also, see if the `exists` clause can be used or not instead of fetching full models.